e2e_playwright/st_text_input_test.py

```diff
@@ -0,0 +1,205 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from playwright.sync_api import Page, expect
+
+from e2e_playwright.conftest import ImageCompareFunction
+
+
+def test_text_input_widget_rendering(
+    themed_app: Page, assert_snapshot: ImageCompareFunction
+):
+    """Test that the st.text_input widgets are correctly rendered via screenshot matching."""
+    text_input_widgets = themed_app.get_by_test_id("stTextInput")
+    expect(text_input_widgets).to_have_count(11)
+
+    assert_snapshot(text_input_widgets.nth(0), name="st_text_input-default")
+    assert_snapshot(text_input_widgets.nth(1), name="st_text_input-value_some_text")
+    assert_snapshot(text_input_widgets.nth(2), name="st_text_input-value_1234")
+    assert_snapshot(text_input_widgets.nth(3), name="st_text_input-value_None")
+    assert_snapshot(text_input_widgets.nth(4), name="st_text_input-placeholder")
+    assert_snapshot(text_input_widgets.nth(5), name="st_text_input-disabled")
+    assert_snapshot(text_input_widgets.nth(6), name="st_text_input-hidden_label")
+    assert_snapshot(text_input_widgets.nth(7), name="st_text_input-collapsed_label")
+    assert_snapshot(text_input_widgets.nth(8), name="st_text_input-callback_help")
+    assert_snapshot(text_input_widgets.nth(9), name="st_text_input-max_chars_5")
+    assert_snapshot(text_input_widgets.nth(10), name="st_text_input-type_password")
+
+
+def test_text_input_has_correct_initial_values(app: Page):
+    """Test that st.text_input has the correct initial values."""
+    markdown_elements = app.get_by_test_id("stMarkdown")
+    expect(markdown_elements).to_have_count(12)
+
+    expected = [
+        "value 1: ",
+        "value 2: some text",
+        "value 3: 1234",
+        "value 4: None",
+        "value 5: ",
+        "value 6: default text",
+        "value 7: default text",
+        "value 8: default text",
+        "value 9: ",
+        "text input changed: False",
+        "value 10: 1234",
+        "value 11: my password",
+    ]
+
+    for markdown_element, expected_text in zip(markdown_elements.all(), expected):
+        expect(markdown_element).to_have_text(expected_text, use_inner_text=True)
+
+
+def test_text_input_shows_instructions_when_dirty(
+    app: Page, assert_snapshot: ImageCompareFunction
+):
+    """Test that st.text_input shows the instructions correctly when dirty."""
+    text_input = app.get_by_test_id("stTextInput").nth(9)
+
+    text_input_field = text_input.locator("input").first
+    text_input_field.fill("123")
+
+    assert_snapshot(text_input, name="st_text_input-input_instructions")
+
+
+def test_text_input_limits_input_via_max_chars(app: Page):
+    """Test that st.text_input correctly limits the number of characters via max_chars."""
+    first_text_input_field = (
```

Comment:
nit: same comment on the variable name here where this isn't the first text input